File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/services/openai/OpenAIDefaults.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.services.openai

import com.microsoft.azure.synapse.ml.param.GlobalParams
import com.microsoft.azure.synapse.ml.services.{OpenAIApiVersion, OpenAISubscriptionKey}
import com.microsoft.azure.synapse.ml.io.http.URLKey
import com.microsoft.azure.synapse.ml.services.aifoundry.AIFoundryModel

object OpenAIDefaults {
  def setDeploymentName(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAIDeploymentNameKey, Left(v))
  }

  def getDeploymentName: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAIDeploymentNameKey))
  }

  def resetDeploymentName(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIDeploymentNameKey)
  }

  def setSubscriptionKey(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAISubscriptionKey, Left(v))
  }

  def getSubscriptionKey: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAISubscriptionKey))
  }

  def resetSubscriptionKey(): Unit = {
    GlobalParams.resetGlobalParam(OpenAISubscriptionKey)
  }

  def setTemperature(v: Double): Unit = {
    require(v >= 0.0 && v <= 2.0, s"Temperature must be between 0.0 and 2.0, got: $v")
    GlobalParams.setGlobalParam(OpenAITemperatureKey, Left(v))
  }

  def getTemperature: Option[Double] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAITemperatureKey))
  }

  def resetTemperature(): Unit = {
    GlobalParams.resetGlobalParam(OpenAITemperatureKey)
  }

  def setURL(v: String): Unit = {
    val url = if (v.endsWith("/")) v else v + "/"
    GlobalParams.setGlobalParam(URLKey, url)
  }

  def getURL: Option[String] = {
    GlobalParams.getGlobalParam(URLKey)
  }

  def resetURL(): Unit = {
    GlobalParams.resetGlobalParam(URLKey)
  }

  def setSeed(v: Int): Unit = {
    GlobalParams.setGlobalParam(OpenAISeedKey, Left(v))
  }

  def getSeed: Option[Int] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAISeedKey))
  }

  def resetSeed(): Unit = {
    GlobalParams.resetGlobalParam(OpenAISeedKey)
  }

  def setTopP(v: Double): Unit = {
    require(v >= 0.0 && v <= 1.0, s"TopP must be between 0.0 and 1.0, got: $v")
    GlobalParams.setGlobalParam(OpenAITopPKey, Left(v))
  }

  def getTopP: Option[Double] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAITopPKey))
  }

  def resetTopP(): Unit = {
    GlobalParams.resetGlobalParam(OpenAITopPKey)
  }

  def setApiVersion(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAIApiVersion, Left(v))
  }

  def getApiVersion: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAIApiVersion))
  }

  def resetApiVersion(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIApiVersion)
  }

  def setModel(v: String): Unit = {
    GlobalParams.setGlobalParam(AIFoundryModel, Left(v))
  }

  def getModel: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(AIFoundryModel))
  }

  def resetModel(): Unit = {
    GlobalParams.resetGlobalParam(AIFoundryModel)
  }

  def setEmbeddingDeploymentName(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAIEmbeddingDeploymentNameKey, Left(v))
  }

  def getEmbeddingDeploymentName: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAIEmbeddingDeploymentNameKey))
  }

  def resetEmbeddingDeploymentName(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIEmbeddingDeploymentNameKey)
  }

  def setVerbosity(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAIVerbosityKey, Left(v))
  }

  def getVerbosity: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAIVerbosityKey))
  }

  def resetVerbosity(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIVerbosityKey)
  }

  def setReasoningEffort(v: String): Unit = {
    GlobalParams.setGlobalParam(OpenAIReasoningEffortKey, Left(v))
  }

  def getReasoningEffort: Option[String] = {
    extractLeft(GlobalParams.getGlobalParam(OpenAIReasoningEffortKey))
  }

  def resetReasoningEffort(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIReasoningEffortKey)
  }

  def setApiType(v: String): Unit = {
    require(
      v == "responses" || v == "chat_completions",
      s"ApiType must be either 'responses' or 'chat_completions', got: $v"
    )
    GlobalParams.setGlobalParam(OpenAIApiTypeKey, v)
  }

  def getApiType: Option[String] = {
    GlobalParams.getGlobalParam(OpenAIApiTypeKey)
  }

  def resetApiType(): Unit = {
    GlobalParams.resetGlobalParam(OpenAIApiTypeKey)
  }

  private def extractLeft[T](optEither: Option[Either[T, String]]): Option[T] = {
    optEither match {
      case Some(Left(v)) => Some(v)
      case _ => None
    }
  }
}

File: cognitive/src/main/python/synapse/ml/services/openai/OpenAIDefaults.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.

import sys

if sys.version >= "3":
    basestring = str

import pyspark
from pyspark import SparkContext


def getOption(opt):
    if opt.isDefined():
        return opt.get()
    else:
        return None


class OpenAIDefaults:
    def __init__(self):
        self.defaults = (
            SparkContext.getOrCreate()._jvm.com.microsoft.azure.synapse.ml.services.openai.OpenAIDefaults
        )

    def set_deployment_name(self, name):
        self.defaults.setDeploymentName(name)

    def get_deployment_name(self):
        return getOption(self.defaults.getDeploymentName())

    def reset_deployment_name(self):
        self.defaults.resetDeploymentName()

    def set_subscription_key(self, key):
        self.defaults.setSubscriptionKey(key)

    def get_subscription_key(self):
        return getOption(self.defaults.getSubscriptionKey())

    def reset_subscription_key(self):
        self.defaults.resetSubscriptionKey()

    def set_temperature(self, temp):
        temp_float = float(temp)
        if not (0.0 <= temp_float <= 2.0):
            raise ValueError(
                f"Temperature must be between 0.0 and 2.0, got: {temp_float}"
            )
        self.defaults.setTemperature(temp_float)

    def get_temperature(self):
        return getOption(self.defaults.getTemperature())

    def reset_temperature(self):
        self.defaults.resetTemperature()

    def set_URL(self, URL):
        self.defaults.setURL(URL)

    def get_URL(self):
        return getOption(self.defaults.getURL())

    def reset_URL(self):
        self.defaults.resetURL()

    def set_seed(self, seed):
        self.defaults.setSeed(int(seed))

    def get_seed(self):
        return getOption(self.defaults.getSeed())

    def reset_seed(self):
        self.defaults.resetSeed()

    def set_top_p(self, top_p):
        top_p_float = float(top_p)
        if not (0.0 <= top_p_float <= 1.0):
            raise ValueError(f"TopP must be between 0.0 and 1.0, got: {top_p_float}")
        self.defaults.setTopP(top_p_float)

    def get_top_p(self):
        return getOption(self.defaults.getTopP())

    def reset_top_p(self):
        self.defaults.resetTopP()

    def set_api_version(self, api_version):
        self.defaults.setApiVersion(api_version)

    def get_api_version(self):
        return getOption(self.defaults.getApiVersion())

    def reset_api_version(self):
        self.defaults.resetApiVersion()

    def set_model(self, ai_foundry_model):
        self.defaults.setModel(ai_foundry_model)

    def get_model(self):
        return getOption(self.defaults.getModel())

    def reset_model(self):
        self.defaults.resetModel()
